src/nbl/video/COpenGLFence.h

```diff
@@ -0,0 +1,78 @@
+#ifndef __NBL_C_OPENGL_FENCE_H_INCLUDED__
+#define __NBL_C_OPENGL_FENCE_H_INCLUDED__
+
+#include "nbl/video/IGPUFence.h"
+#include "nbl/video/IOpenGLSyncPrimitiveBase.h"
+#include "nbl/video/IOpenGL_FunctionTable.h"
+
+namespace nbl {
+namespace video
+{
+
+class COpenGLFence final : public IGPUFence, public IOpenGLSyncPrimitiveBase
+{
+protected:
+    ~COpenGLFence() = default;
+
+public:
+    // signaled ctor
+    COpenGLFence(IOpenGL_LogicalDevice* dev, ILogicalDevice* _dev, IOpenGL_FunctionTable* gl) : IGPUFence(_dev, ECF_SIGNALED_BIT), IOpenGLSyncPrimitiveBase(dev)
+    {
+        signal(gl);
+    }
+    // un-signaled ctor
+    explicit COpenGLFence(IOpenGL_LogicalDevice* dev, ILogicalDevice* _dev) : IGPUFence(_dev, static_cast<E_CREATE_FLAGS>(0)), IOpenGLSyncPrimitiveBase(dev)
+    {
+
+    }
+
+    E_STATUS wait(IOpenGL_FunctionTable* _gl, uint64_t timeout)
+    {
+        const uint64_t pretimeout = prewait();
+
+        // TODO
+        // im not sure if this is proper vulkan emulation
+        // @matt ?
+        {
+            if (pretimeout >= timeout)
+                return ES_TIMEOUT;
+            timeout -= pretimeout;
+        }
+
+        COpenGLSync::E_STATUS status = m_sync->waitCPU(_gl, timeout);
+        if (status == COpenGLSync::ES_FAIL)
+            return ES_ERROR;
+        else if (status == COpenGLSync::ES_TIMEOUT_EXPIRED)
+            return ES_TIMEOUT;
+        else
+            return ES_SUCCESS;
+    }
+
+    E_STATUS getStatus(IOpenGL_FunctionTable* _gl)
+    {
+        if (!isWaitable())
+            return ES_NOT_READY;
```

Comment:
if properly implemented, the `COpenGLSync::waitCPU` should return a timeout if the `GLsync` is a nullptr